PhysicsTools/PatUtils/plugins/L1PrefiringWeightProducer.cc

```diff
@@ -0,0 +1,478 @@
+
+  TF1* parametrization0p0To0p2_;
+  TF1* parametrization0p2To0p3_;
+  TF1* parametrization0p3To0p55_;
+  TF1* parametrization0p55To0p83_;
+  TF1* parametrization0p83To1p24_;
+  TF1* parametrization1p24To1p4_;
+  TF1* parametrization1p4To1p6_;
+  TF1* parametrization1p6To1p8_;
+  TF1* parametrization1p8To2p1_;
+  TF1* parametrization2p1To2p25_;
+  TF1* parametrization2p25To2p4_;
+  TF1* parametrizationHotSpot_;
+
+  TH2F* h_prefmap_photon;
+  TH2F* h_prefmap_jet;
+  std::string dataera_;
+  std::string dataeraMuon_;
+  bool useEMpt_;
+  double prefiringRateSystUnc_;
+  double jetMaxMuonFraction_;
+  bool skipwarnings_;
+};
+
+L1PrefiringWeightProducer::L1PrefiringWeightProducer(const edm::ParameterSet& iConfig) {
+  photons_token_ = consumes<std::vector<pat::Photon> >(iConfig.getParameter<edm::InputTag>("ThePhotons"));
+  jets_token_ = consumes<std::vector<pat::Jet> >(iConfig.getParameter<edm::InputTag>("TheJets"));
+  muon_token_ = consumes<std::vector<pat::Muon> >(iConfig.getParameter<edm::InputTag>("TheMuons"));
+
+  dataera_ = iConfig.getParameter<std::string>("DataEraECAL");
+  dataeraMuon_ = iConfig.getParameter<std::string>("DataEraMuon");
+  useEMpt_ = iConfig.getParameter<bool>("UseJetEMPt");
+  prefiringRateSystUnc_ = iConfig.getParameter<double>("PrefiringRateSystematicUncty");
+  jetMaxMuonFraction_ = iConfig.getParameter<double>("JetMaxMuonFraction");
+  skipwarnings_ = iConfig.getParameter<bool>("SkipWarnings");
+
+  TFile* file_prefiringmaps_;
+  std::string fname = iConfig.getParameter<std::string>("L1Maps");
+  edm::FileInPath mapsfilepath("PhysicsTools/PatUtils/data/" + fname);
+  file_prefiringmaps_ = new TFile(mapsfilepath.fullPath().c_str(), "read");
+  if (file_prefiringmaps_ == nullptr && !skipwarnings_)
+    edm::LogWarning("L1PrefireWeightProducer")
+        << "File with maps not found. All prefiring weights set to 0. " << std::endl;
+
+  TString mapphotonfullname = "L1prefiring_photonptvseta_" + dataera_;
+  if (!file_prefiringmaps_->Get(mapphotonfullname) && !skipwarnings_)
+    edm::LogWarning("L1PrefireWeightProducer")
+        << "Photon map not found. All photons prefiring weights set to 0. " << std::endl;
+  h_prefmap_photon = (TH2F*)file_prefiringmaps_->Get(mapphotonfullname);
+
+  TString mapjetfullname = (useEMpt_) ? "L1prefiring_jetemptvseta_" + dataera_ : "L1prefiring_jetptvseta_" + dataera_;
+  if (!file_prefiringmaps_->Get(mapjetfullname) && !skipwarnings_)
+    edm::LogWarning("L1PrefireWeightProducer")
+        << "Jet map not found. All jets prefiring weights set to 0. " << std::endl;
+  h_prefmap_jet = (TH2F*)file_prefiringmaps_->Get(mapjetfullname);
+  file_prefiringmaps_->Close();
+  delete file_prefiringmaps_;
+
+  TFile* file_prefiringparams_;
+  std::string fnameMuon = iConfig.getParameter<std::string>("L1MuonParametrizations");
+  edm::FileInPath paramsfilepath("PhysicsTools/PatUtils/data/" + fnameMuon);
+  file_prefiringparams_ = new TFile(paramsfilepath.fullPath().c_str(), "read");
```

Comment:
This is a memory leak. Make this a member data instead and use a `std::unique_ptr<TFile>` instead. I’m pretty sure that all the item that were retrieved from calling Get are owned by this TFile and will be deleted by it. Therefore you need the lifetime of this TFile to be for the whole job.

Reply:
Fixed.